src/presets.js

```diff
@@ -169,6 +198,30 @@ const attachNestedChangelog = (upstreams, commit, callback) => {
 					cb,
 				);
 			}
+			const regexp1 = new RegExp(`[Uu]pdate (layers)?/${upstream.pattern}`);
+			const match1 = commit.body.match(regexp1);
+			if (match1) {
+				const result = childProcess.spawnSync('git', [
```

Comment:
Maybe from here and further down, it could be in its own function.

Reply:
thanks @stathismor you're right that `layers` is balena specific.

Not sure how to approach this from `versionist.conf.js`, could you maybe explain your thoughts?

Another way to approach it would be to modify the way the upstream repositories are defined in `repo.yml` as currently we only state the submodule name and not the path, as in:
```
upstream:                                                                                               
- repo: 'meta-balena'                                                         
    url: 'http://github.com/balena-os/meta-balena'                              
- repo: 'balena-yocto-scripts'                                                
    url: 'http://github.com/balena-os/balena-yocto-scripts'
```
Instead of:
```
- repo: 'layers/meta-balena'                                                         
    url: 'http://github.com/balena-os/meta-balena'    
```
This would involve changing all the current `repo.yml`, and I am not sure if that change would affect anything else.

Finally, another solution not that intrusive is using a broader regex and still relying on the submodule name for identification:
```
const regexp1 = new RegExp(`[Uu]pdate (.*)?${upstream.pattern}`);
```
What do you think?

Reply:
1.
> Not sure how to approach this from versionist.conf.js, could you maybe explain your thoughts?

Yeah, sure. You know that repos can have [custom versionist configs](https://github.com/product-os/versionist#configuration), right? An example how that can look [here](https://github.com/product-os/balena-versionist/blob/master/lib/repo-type-mappings/yocto-layer/versionist.conf.js). A `versionist.conf.js` at the repo's root, would overwrite any default `versionist`/`balena-versionist` behaviour. Looking at it now, the relevant option here I think is `transformTemplateDataAsync`. Current code does:
```
	transformTemplateDataAsync: {
		'nested-changelogs': (options, data, callback) => {
			async.map(
				data.commits,
				_.partial(attachNestedChangelog, options.upstream),
				(err, commits) => {
					if (err) {
						return callback(err);
					}
					data.commits = _.sortBy(commits, (commit) => {
						return Boolean(commit.nested);
					});

					return callback(err, data);
				},
			);
		},
	},
```
You could copy that on the conf file, along with your own implementation of `attachNestedChangelog`, as you have here.

If it's just one or two repos that want this behavior, the conf file would be on their root. If we want to have this for all repos of a type, it could go to `balena-versionist` configurations https://github.com/product-os/balena-versionist/tree/master/lib/repo-type-mappings

2. Regarding changing upstream names or having a more generic regexp. We're still using `repo.yml` though which is a balena thing... Also, wouldn't having a generic regexp increase the chances of confusing the tool, when someone just writes `Update <submodule name>`?

Reply:
thanks @stathismor I understand now how to approach this from `balena-versionist` and `versionist.conf.js`.

However, the reason why I decided to put this into versionist initially is because it is a usage improvement. Right now you have to manually craft a commit body to link the nested changelog behaviour, and you have to manually type the current and new versions, and you can get it wrong and there is no way back. But versionist has all the information to extract those versions for you, you only have to match the submodule name in the commit log with the one defined in `repo.yml`.

Not sure that `repo.yml` is a balena thing. I don't think you can use nested changelogs if you don't define the upstreams in `repo.yml`, so we either agree that nested changelogs is balena-specific, or `repo.yml` is not.

In summary,

* I can refactor the change into `versionist.conf.js, making it balena specific
* We can either change the upstreams in `repo.yml` or extend the regex.

About extending the regex, not sure what you mean by:

> confusing the tool, when someone just writes Update <submodule name>

The tool will try to extract nested changelogs in that case, but if there are no upstreams defined in `repo.yml` or not submodule data, it will not find nested data and return empty.

Anyway, I really just want to solve this for device repositories, so whatever you guys decide is the path forward is fine with me.

Reply:
I think I understand why a more generic regex is messy, so it's between moving it to `versionist.conf.js` or modifying the upstreams in `repo.yml`.

Thinking about it, I think that it's a usage improvement on the nested changelog functionality already in `versionist` so I'd like to understand what are the drawbacks in changing the upstreams in `repo.yml`

So, I am tending to try the following:
Leave the regex match as:
```
const regexp1 = new RegExp(`[Uu]pdate ${upstream.pattern}`);
```
And change the `repo.yml` in device repositories to match as in:
```
- repo: 'layers/meta-balena'                                                         
    url: 'http://github.com/balena-os/meta-balena'
- repo: 'balena-yocto-scripts'                                                
    url: 'http://github.com/balena-os/balena-yocto-scripts'
```
What do you think?

Reply:
> Not sure that repo.yml is a balena thing. I don't think you can use nested changelogs if you don't define the upstreams in repo.yml, so we either agree that nested changelogs is balena-specific, or repo.yml is not

Fair point, I missed that nested changelogs do use `repo.yml` anyway.

Ok, I am fine with changing upstreams :+1:

Reply:
I have now removed `layers` and just use the defined `upstream` from `repo.yml`. I have updated my test setup and been testing with a few submodule PRs, all looks good.
I expect that most projects are using submodules at the project root and not in a submodule, so only the device repositories will need to update `repo.yml`.